java-frontend/src/main/java/org/sonar/java/se/checks/DivisionByZeroCheck.java

```diff
@@ -0,0 +1,299 @@
+  private enum Status {
+    ZERO, NON_ZERO, UNDETERMINED;
+  }
+
+  private static class ZeroSymbolicValue extends SymbolicValue {
+
+    private final Status initialStatus;
+
+    public ZeroSymbolicValue(int id, Status initialConstraint) {
+      super(id);
+      this.initialStatus = initialConstraint;
+    }
+
+  }
+
+  private static class ZeroConstraint extends ObjectConstraint {
+    private ZeroConstraint(Tree syntaxNode, Status status) {
+      super(false, false, syntaxNode, status);
+    }
+
+    @Override
+    public ZeroConstraint inverse() {
+      return new ZeroConstraint(syntaxNode(), Status.NON_ZERO);
+    }
+
+    @Override
+    public boolean isInversible() {
+      return hasStatus(Status.ZERO);
+    }
+
+    @Override
+    public String toString() {
+      // For debugging purpose
+      return "" + (hasStatus(Status.ZERO) ? Status.ZERO : (hasStatus(Status.NON_ZERO) ? Status.NON_ZERO : Status.UNDETERMINED));
+    }
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    PreStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private class PreStatementVisitor extends CheckerTreeNodeVisitor {
+    private final ConstraintManager constraintManager;
+    private final CheckerContext context;
+
+    PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+      this.context = context;
+      this.constraintManager = context.getConstraintManager();
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      List<SymbolicValue> symbolicValues = programState.peekValues(2);
+      SymbolicValue var = symbolicValues.get(0);
+      SymbolicValue expr = symbolicValues.get(1);
+      checkOperation(tree, var, expr);
+    }
+
+    @Override
+    public void visitBinaryExpression(BinaryExpressionTree tree) {
+      List<SymbolicValue> symbolicValues;
+      switch (tree.kind()) {
+        case MULTIPLY:
+        case PLUS:
+        case MINUS:
+        case DIVIDE:
+        case REMAINDER:
+          symbolicValues = programState.peekValues(2);
+          checkOperation(tree, symbolicValues.get(1), symbolicValues.get(0));
+          break;
+        case GREATER_THAN:
+        case GREATER_THAN_OR_EQUAL_TO:
+        case LESS_THAN:
+        case LESS_THAN_OR_EQUAL_TO:
+          symbolicValues = programState.peekValues(2);
+          addUnderterminedConstraint(symbolicValues.get(1), tree.leftOperand());
+          addUnderterminedConstraint(symbolicValues.get(0), tree.rightOperand());
+          break;
+        default:
+          // do nothing
+      }
+    }
+
+    private void addUnderterminedConstraint(SymbolicValue sv, Tree tree) {
+      programState = programState.addConstraint(sv, new ZeroConstraint(tree, Status.UNDETERMINED));
+    }
+
+    private void checkOperation(Tree tree, SymbolicValue leftOp, SymbolicValue rightOp) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "checkOperation" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)